p2p/memp2p/memp2p.go

```diff
@@ -0,0 +1,288 @@
+package memp2p
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/p2p"
+)
+
+type MemP2PMessenger struct {
+	Network     *MemP2PNetwork
+	P2P_ID      p2p.PeerID
+	Address     string
+	Topics      map[string]p2p.MessageProcessor
+	TopicsMutex sync.RWMutex
+}
+
+func NewMemP2PMessenger(network *MemP2PNetwork) *MemP2PMessenger {
+	ID := fmt.Sprintf("Peer%d", len(network.PeerIDs())+1)
+	Address := fmt.Sprintf("/memp2p/%s", string(ID))
+
+	messenger := &MemP2PMessenger{
+		Network:     network,
+		P2P_ID:      p2p.PeerID(ID),
+		Address:     Address,
+		Topics:      make(map[string]p2p.MessageProcessor),
+		TopicsMutex: sync.RWMutex{},
+	}
+
+	network.RegisterPeer(messenger)
+
+	return messenger
+}
+
+func (messenger *MemP2PMessenger) ID() p2p.PeerID {
+	return messenger.P2P_ID
+}
+
+func (messenger *MemP2PMessenger) Peers() []p2p.PeerID {
+	fmt.Printf("debug:Peers\n")
+	// If the messenger is connected to the network, it has knowledge of all
+	// other peers.
+	if !messenger.IsConnectedToNetwork() {
+		return []p2p.PeerID{}
+	}
+	return messenger.Network.PeerIDs()
+}
+
+func (messenger *MemP2PMessenger) Addresses() []string {
+	addresses := make([]string, 1)
+	addresses[0] = messenger.Address
+	return addresses
+}
+
+func (messenger *MemP2PMessenger) ConnectToPeer(address string) error {
+	fmt.Printf("debug:ConnectToPeer\n")
+	if !messenger.IsConnectedToNetwork() {
+		return errors.New("Peer not connected to network, can't connect to any other peer.")
+	}
+	// Do nothing, all peers are connected to each other already.
+	return nil
+}
+
+func (messenger *MemP2PMessenger) IsConnectedToNetwork() bool {
+	fmt.Printf("debug:IsConnectedToNetwork\n")
+	return messenger.Network.IsPeerConnected(messenger.ID())
+}
+
+func (messenger *MemP2PMessenger) IsConnected(peerID p2p.PeerID) bool {
+	fmt.Printf("debug:IsConnected\n")
+	// If the messenger is connected to the network, it is connected to all other peers.
+	return messenger.IsConnectedToNetwork()
+}
+
+func (messenger *MemP2PMessenger) ConnectedPeers() []p2p.PeerID {
+	fmt.Printf("debug:ConnectedPeers\n")
+	if !messenger.IsConnectedToNetwork() {
+		return []p2p.PeerID{}
+	}
+	return messenger.Network.PeerIDsExceptOne(messenger.ID())
+}
+
+func (messenger *MemP2PMessenger) ConnectedAddresses() []string {
+	fmt.Printf("debug:ConnectedAddresses\n")
+	if !messenger.IsConnectedToNetwork() {
+		return []string{}
+	}
+	return messenger.Network.ListAddressesExceptOne(messenger.ID())
+}
+
+func (messenger *MemP2PMessenger) PeerAddress(pid p2p.PeerID) string {
+	return fmt.Sprintf("/memp2p/%s", string(pid))
+}
+
+func (messenger *MemP2PMessenger) ConnectedPeersOnTopic(topic string) []p2p.PeerID {
+	fmt.Printf("debug:ConnectedPeersOnTopic\n")
+	var filteredPeers []p2p.PeerID
+	if !messenger.IsConnectedToNetwork() {
+		return filteredPeers
+	}
+
+	messenger.TopicsMutex.Lock()
```

Comment:
`messenger.TopicsMutex.Lock()` is called here to prevent the deletion of the topic while iterating over the peers of `messenger`. Below, the call to `HasTopic` is on the peers of the `messenger`, which involves different mutex instances. At least this was my intention. Is the lock unnecessary?

Reply:
I was confusing something. The lock is indeed unnecessary. Removed it.